internal/rules/mechanisms/authenticators/jwt_authenticator.go

```diff
@@ -91,16 +100,24 @@ func newJwtAuthenticator(id string, rawConfig map[string]any) (*jwtAuthenticator
 		return nil, err
 	}
 
-	if conf.Endpoint.Headers == nil {
-		conf.Endpoint.Headers = make(map[string]string)
+	// Pick the right endpoint to apply defaults to, JWKSEndpoint if set, otherwise the discovery endpoint
+	endpoint := x.IfThenElse(conf.JWKSEndpoint != nil, conf.JWKSEndpoint, conf.DiscoveryEndpoint)
```

Comment:
Without being involved into the implementation...
Would it not be easier to introduce an interface, e.g. JWKSFetcher or alike and have two implementation for that. The one would do, what was previously there and the second would wrap the first and cope with the discovery document?
I think that approach would reduce the amount of code and make it also easier to test.
But this is just an idea.

Reply:
I did think about that, problem with that approach is, that the discovery document retrieves more values than just the JWKS, so an interface like this would need to also cover handling of issuers and algorithms, I think. 

Right now it's implemented in a way that fetchJWKS also returns an oauth2.expectation which gets merged with the user specific one, but I'm not sure how smart that is.

Reply:
Yeah. That "thing" with the "expectations" feels a bit strange. 
What about the following?: emulating the discovery endpoint if only the jwks endpoint is configured. So you would always internally call the method responsible for fetching the discovery document. It would always contain the jwks endpoint if only that was configured and the other properties otherwise. Then you would just use the discovery document to actually fetch the jwks and also to merge the "expectations" into the configured ones.

Reply:
What I mean here is chaning the `verifyToken` as follows.
```go
func (a *jwtAuthenticator) verifyToken(ctx heimdall.Context, token *jwt.JSONWebToken) (json.RawMessage, error) {
        doc, err := fetchDiscoveryDocument(ctx, token)
        // error handling here

	if len(token.Headers[0].KeyID) == 0 {
		return a.verifyTokenWithoutKID(ctx, doc, token)
	}

	sigKey, err := a.getKey(ctx, doc, token.Headers[0].KeyID)
	if err != nil {
		return nil, err
	}

	return a.verifyTokenWithKey(token, doc, sigKey)
}
```

in this context the `doc` would have the JWKS property or alike being the url to the jwks endpoint and the other settings, which would be used to assert the JWT together with the optionally configured assertions